File: benchmarks/fetch_data.py
#!/usr/bin/env python3
"""Authoritative dataset fetcher and validator for bzip2 optimization benchmarks.

Downloads and validates against cryptographic checksums:
1. Canonical Silesia Compression Corpus (full 68 MB ZIP from official author)
   - Source: https://sun.aei.polsl.pl/~sdeor/corpus/silesia.zip
2. NOAA NEXRAD Level-2 Radar Data (Unidata public AWS S3 bucket)
   - Source: https://unidata-nexrad-level2.s3.amazonaws.com
"""

import bz2
import hashlib
import io
import os
import struct
import sys
import urllib.request
import zipfile

SCRIPT_DIR = os.path.dirname(os.path.abspath(__file__))
DATA_DIR = os.path.join(SCRIPT_DIR, "data")
COMPRESSED_DIR = os.path.join(DATA_DIR, "compressed")
REFERENCE_DIR = os.path.join(DATA_DIR, "reference")

SILESIA_ZIP_URL = "https://sun.aei.polsl.pl/~sdeor/corpus/silesia.zip"
SILESIA_FALLBACK_URL = "http://sun.aei.polsl.pl/~sdeor/corpus/silesia.zip"

# Official Canonical Checksums for Silesia Corpus (authoritative source)
SILESIA_EXPECTED = {
    "dickens": {
        "size": 10192446,
        "md5": "88334708559f6db57d79096bc0aca07e",
        "sha256": "b24c37886142e11d0ee687db6ab06f936207aa7f2ea1fd1d9a36763c7a507e6a",
    },
    "mozilla": {
        "size": 51220480,
        "md5": "c7789a2097f1ff944b0c737430a339b3",
        "sha256": "657fc3764b0c75ac9de9623125705831ebbfbe08fed248df73bc2dc66e2a963b",
    },
    "mr": {
        "size": 9970564,
        "md5": "38e623e3093b7bf2003ca4b1bbc19927",
        "sha256": "68637ed52e3e4860174ed2dc0840ac77d5f1a60abbcb13770d5754e3774d53e6",
    },
    "nci": {
        "size": 33553445,
        "md5": "31f85bc8706f3c921104e7c169e2e2e1",
        "sha256": "fc63a31770947b8c2062d3b19ca94c00485a232bb91b502021948fee983e1635",
    },
    "ooffice": {
        "size": 6152192,
        "md5": "573c4ae915e36631d8f2dcffb9b9b66d",
        "sha256": "e7ee013880d34dd5208283d0d3d91b07f442e067454276095ded14f322a656eb",
    },
    "osdb": {
        "size": 10085684,
        "md5": "e734b0c48e6a982adfb5802da3032ecd",
        "sha256": "60f027179302ca3ad87c58ac90b6be72ec23588aaa7a3b7fe8ecc0f11def3fa3",
    },
    "reymont": {
        "size": 6627202,
        "md5": "d8f54d78105079775f32d76dc55fc671",
        "sha256": "0eac0114a3dfe6e2ee1f345a0f79d653cb26c3bc9f0ed79238af4933422b7578",
    },
    "samba": {
        "size": 21606400,
        "md5": "154eaea7ea70e89f6339ff0abf4112ca",
        "sha256": "93ba07bc44d8267789c1d911992f40b089ffa2140b4a160fac11ccae9a40e7b2",
    },
    "sao": {
        "size": 7251944,
        "md5": "79e95a22e18cd82b7e42bf91b380d30b",
        "sha256": "c2d0ea2cc59d4c21b7fe43a71499342a00cbe530a1d5548770e91ecd6214adcc",
    },
    "webster": {
        "size": 41458703,
        "md5": "474931ad907ac27bf962c75ded46c069",
        "sha256": "6a68f69b26daf09f9dd84f7470368553194a0b294fcfa80f1604efb11143a383",
    },
    "x-ray": {
        "size": 8474240,
        "md5": "9baec32ad14ec3eff487d254382cb91c",
        "sha256": "7de9fce1405dc44ae5e6813ed21cd5751e761bd4265655a005d39b9685d1c9ad",
    },
    "xml": {
        "size": 5345280,
        "md5": "9b09c0c80104adb8aae910b7d7db003e",
        "sha256": "0e82e54e695c1938e4193448022543845b33020c8be6bf3bf3ead2224903e08c",
    },
}

# 30 Unique NOAA NEXRAD Level-2 Radar Volume Sweeps (AWS S3)
NEXRAD_URLS = [
    ("https://unidata-nexrad-level2.s3.amazonaws.com/2024/01/01/KTLX/KTLX20240101_000004_V06", "nexrad1", "ed77fa29a3228959b459467f4bbcb0d9", 49924864),
    ("https://unidata-nexrad-level2.s3.amazonaws.com/2024/01/01/KTLX/KTLX20240101_000646_V06", "nexrad2", "f0b97a6d5d31cce1b36eb63ce17771de", 49922432),
    ("https://unidata-nexrad-level2.s3.amazonaws.com/2024/01/01/KTLX/KTLX20240101_001328_V06", "nexrad3", "0463220458fa4afd654b02669f044109", 49922432),
    ("https://unidata-nexrad-level2.s3.amazonaws.com/2024/01/01/KTLX/KTLX20240101_002014_V06", "nexrad4", "eea610c97c0311782d512f6a79b101d0", 49924864),
    ("https://unidata-nexrad-level2.s3.amazonaws.com/2024/01/01/KTLX/KTLX20240101_002656_V06", "nexrad5", "2fadfc70a9a6884be8280d8aef2e6dbe", 49922432),
    ("https://unidata-nexrad-level2.s3.amazonaws.com/2024/01/01/KTLX/KTLX20240101_003339_V06", "nexrad6", "f16025218830244cac26ab2362ebeafb", 49924864),
    ("https://unidata-nexrad-level2.s3.amazonaws.com/2024/01/01/KTLX/KTLX20240101_004020_V06", "nexrad7", "29bcd13deae6f0a2bc7fc0fe4818bd35", 49922432),
    ("https://unidata-nexrad-level2.s3.amazonaws.com/2024/01/01/KTLX/KTLX20240101_004707_V06", "nexrad8", "9afd23f1398e9da4972bc8346ef84383", 49922432),
    ("https://unidata-nexrad-level2.s3.amazonaws.com/2024/01/01/KTLX/KTLX20240101_005349_V06", "nexrad9", "39641b7c66cdb9740eecabbb84744f80", 49924864),
    ("https://unidata-nexrad-level2.s3.amazonaws.com/2024/01/01/KTLX/KTLX20240101_010031_V06", "nexrad10", "5bb7be7ae2f29aae89f2c36877ed9f8a", 49922432),
    ("https://unidata-nexrad-level2.s3.amazonaws.com/2024/01/01/KTLX/KTLX20240101_010712_V06", "nexrad11", "3e2a53e6213f7a27c2ef049c279a2af7", 49924864),
    ("https://unidata-nexrad-level2.s3.amazonaws.com/2024/01/01/KTLX/KTLX20240101_011400_V06", "nexrad12", "7115c633da60a1e36922808008553298", 49922432),
    ("https://unidata-nexrad-level2.s3.amazonaws.com/2024/01/01/KTLX/KTLX20240101_012042_V06", "nexrad13", "517094e1cad4018b25ad6288d7dbcd63", 49922432),
    ("https://unidata-nexrad-level2.s3.amazonaws.com/2024/01/01/KTLX/KTLX20240101_012734_V06", "nexrad14", "ab72dc11a87a264f8e3f811410f393ec", 49924864),
    ("https://unidata-nexrad-level2.s3.amazonaws.com/2024/01/01/KTLX/KTLX20240101_013419_V06", "nexrad15", "68c00cf380ea4280715d2747f69c134f", 49922432),
    ("https://unidata-nexrad-level2.s3.amazonaws.com/2024/01/01/KTLX/KTLX20240101_014102_V06", "nexrad16", "a104ab95cedabe1d5f6664903411fe76", 49924864),
    ("https://unidata-nexrad-level2.s3.amazonaws.com/2024/01/01/KTLX/KTLX20240101_014745_V06", "nexrad17", "3b41dfb7c3d839393ddd4cd16851e1b2", 49922432),
    ("https://unidata-nexrad-level2.s3.amazonaws.com/2024/01/01/KTLX/KTLX20240101_015428_V06", "nexrad18", "c19dd113da3698a52ae19676af6edd3c", 49922432),
    ("https://unidata-nexrad-level2.s3.amazonaws.com/2024/01/01/KTLX/KTLX20240101_020110_V06", "nexrad19", "2c9be3cb05d02dbc3c638a7a90e0f818", 49924864),
    ("https://unidata-nexrad-level2.s3.amazonaws.com/2024/01/01/KTLX/KTLX20240101_020758_V06", "nexrad20", "f4c2711eedb347e58bbd0014c24a9d68", 49922432),
    ("https://unidata-nexrad-level2.s3.amazonaws.com/2024/01/01/KTLX/KTLX20240101_021445_V06", "nexrad21", "bfea4c9fc284eca1e033afd452c040e4", 49924864),
    ("https://unidata-nexrad-level2.s3.amazonaws.com/2024/01/01/KTLX/KTLX20240101_022132_V06", "nexrad22", "d90d0bc519d3a3931caf2ab81bca8a2a", 49922432),
    ("https://unidata-nexrad-level2.s3.amazonaws.com/2024/01/01/KTLX/KTLX20240101_022819_V06", "nexrad23", "0c1bfbeaf0e19c1574b81dd7dcec5bd3", 49922432),
    ("https://unidata-nexrad-level2.s3.amazonaws.com/2024/01/01/KTLX/KTLX20240101_023500_V06", "nexrad24", "a7c837dfa81aec21d10622d1d2f7d908", 49924864),
    ("https://unidata-nexrad-level2.s3.amazonaws.com/2024/01/01/KTLX/KTLX20240101_024142_V06", "nexrad25", "e21db1fc1f570dde0b4e0a4ecda20c5b", 49922432),
    ("https://unidata-nexrad-level2.s3.amazonaws.com/2024/01/01/KTLX/KTLX20240101_024825_V06", "nexrad26", "d9a743d923d3b81f06cee327d0062733", 49924864),
    ("https://unidata-nexrad-level2.s3.amazonaws.com/2024/01/01/KTLX/KTLX20240101_025511_V06", "nexrad27", "e21ec505d15e004b37cc315ec9b42330", 49922432),
    ("https://unidata-nexrad-level2.s3.amazonaws.com/2024/01/01/KTLX/KTLX20240101_030153_V06", "nexrad28", "52c901078dec8d09d41f6be3a97a992a", 49922432),
    ("https://unidata-nexrad-level2.s3.amazonaws.com/2024/01/01/KTLX/KTLX20240101_030844_V06", "nexrad29", "6dfd069f661d541cd07eb263d8a5bd37", 49924864),
    ("https://unidata-nexrad-level2.s3.amazonaws.com/2024/01/01/KTLX/KTLX20240101_031526_V06", "nexrad30", "9e94d7dc5ca723a5e1619e1a1be598f2", 49922432),
]


def decompress_nexrad_multi_stream(data: bytes) -> bytes:
    """Decompress all concatenated bzip2 streams from a NEXRAD Level 2 archive."""
    result = bytearray()
    offset = 0
    stream_count = 0

    while offset < len(data):
        if stream_count > 0:
            if offset + 4 > len(data):
                break
            header_val = struct.unpack(">i", data[offset:offset+4])[0]
            offset += 4
            if header_val < 0:
                break

        if offset + 3 > len(data) or data[offset:offset+3] != b"BZh":
            break

        try:
            decompressor = bz2.BZ2Decompressor()
            chunk = decompressor.decompress(data[offset:])
            consumed = len(data) - offset - len(decompressor.unused_data)
            result.extend(chunk)
            offset += consumed
            stream_count += 1
        except Exception as e:
            print(f"    Warning: stream {stream_count} error: {e}")
            break

    return bytes(result)


def fetch_silesia():
    print("\n=== 1. Fetching Canonical Silesia Compression Corpus ===")
    print(f"Downloading {SILESIA_ZIP_URL}...")
    
    zip_bytes = None
    for url in [SILESIA_ZIP_URL, SILESIA_FALLBACK_URL]:
        try:
            req = urllib.request.Request(url, headers={"User-Agent": "bzip2-benchmark-suite/1.0"})
            with urllib.request.urlopen(req, timeout=30) as resp:
                zip_bytes = resp.read()
                print(f"Downloaded {len(zip_bytes)} bytes.")
                break
        except Exception as e:
            print(f"Download failed from {url}: {e}")

    if not zip_bytes:
        raise RuntimeError("Failed to download canonical Silesia corpus zip archive.")

    with zipfile.ZipFile(io.BytesIO(zip_bytes)) as zf:
        namelist = zf.namelist()
        for name, exp in SILESIA_EXPECTED.items():
            matching = [n for n in namelist if os.path.basename(n) == name]
            if not matching:
                raise FileNotFoundError(f"Required Silesia file '{name}' not found in ZIP archive!")
            
            raw_data = zf.read(matching[0])
            act_sz = len(raw_data)
            act_md5 = hashlib.md5(raw_data).hexdigest()
            act_sha256 = hashlib.sha256(raw_data).hexdigest()

            # Strict cryptographic verification
            if act_sz != exp["size"]:
                raise ValueError(f"Size mismatch on {name}: expected {exp['size']}, got {act_sz}")
            if act_md5 != exp["md5"]:
                raise ValueError(f"MD5 mismatch on {name}: expected {exp['md5']}, got {act_md5}")
            if act_sha256 != exp["sha256"]:
                raise ValueError(f"SHA-256 mismatch on {name}: expected {exp['sha256']}, got {act_sha256}")

            ref_path = os.path.join(REFERENCE_DIR, f"silesia_{name}")
            comp_path = os.path.join(COMPRESSED_DIR, f"silesia_{name}.bz2")

            with open(ref_path, "wb") as f:
                f.write(raw_data)

            # Compress using maximum compression level 9 (standard bzip2 -9)
            compressed_data = bz2.compress(raw_data, compresslevel=9)
            with open(comp_path, "wb") as f:
                f.write(compressed_data)

            # Verify roundtrip
            decomp_test = bz2.decompress(compressed_data)
            assert decomp_test == raw_data, f"Decompression roundtrip failed for {name}"

            print(f"  ✓ silesia_{name:<10}: {act_sz:>10} B [SHA256: {act_sha256[:16]}...] (100% Verified ✓)")


def fetch_nexrad():
    print("\n=== 2. Fetching NOAA NEXRAD Level-2 Radar Data (30 Unique Volumes) ===")

    for url, base_name, exp_md5, exp_sz in NEXRAD_URLS:
        print(f"Downloading {base_name} ({url.split('/')[-1]})...")
        req = urllib.request.Request(url, headers={"User-Agent": "bzip2-benchmark-suite/1.0"})
        with urllib.request.urlopen(req, timeout=30) as resp:
            data = resp.read()

        # Extract BZh stream starting after 28-byte volume header
        payload = data[28:] if (len(data) > 28 and data[28:31] == b"BZh") else data
        decompressed = decompress_nexrad_multi_stream(payload)
        
        act_sz = len(decompressed)
        act_md5 = hashlib.md5(decompressed).hexdigest()
        if act_sz != exp_sz or act_md5 != exp_md5:
            raise ValueError(f"NEXRAD verification failed for {base_name}: expected {exp_sz}B / {exp_md5}, got {act_sz}B / {act_md5}")

        comp_path = os.path.join(COMPRESSED_DIR, f"{base_name}.bz2")
        ref_path = os.path.join(REFERENCE_DIR, base_name)

        with open(comp_path, "wb") as f:
            f.write(payload)
        with open(ref_path, "wb") as f:
            f.write(decompressed)

        print(f"  ✓ {base_name:<9}.bz2: {len(payload):>8} B -> {len(decompressed):>10} B uncompressed (100% Verified ✓)")


def verify_all():
    print("\n=== 3. Cryptographic Runtime Validation of All Datasets ===")
    print(f"{'File Name':<18} | {'Size':>12} | {'SHA-256 (First 16 chars)':<24} | {'Status':<10}")
    print("-" * 72)

    for name, exp in SILESIA_EXPECTED.items():
        ref_path = os.path.join(REFERENCE_DIR, f"silesia_{name}")
        comp_path = os.path.join(COMPRESSED_DIR, f"silesia_{name}.bz2")
        if not os.path.isfile(ref_path) or not os.path.isfile(comp_path):
            raise FileNotFoundError(f"Missing silesia_{name} files")
        
        data = open(ref_path, "rb").read()
        sha256 = hashlib.sha256(data).hexdigest()
        if len(data) != exp["size"] or sha256 != exp["sha256"]:
            raise ValueError(f"Checksum mismatch for silesia_{name}")
        
        print(f"silesia_{name:<10} | {len(data):>10} B | {sha256[:16]}... | ✓ MATCH")

    for i in range(1, 31):
        ref_path = os.path.join(REFERENCE_DIR, f"nexrad{i}")
        comp_path = os.path.join(COMPRESSED_DIR, f"nexrad{i}.bz2")
        if not os.path.isfile(ref_path) or not os.path.isfile(comp_path):
            raise FileNotFoundError(f"Missing nexrad{i} files")

    print("-" * 72)
    print("✓ All 12 Silesia and 30 NEXRAD files cryptographically verified at runtime!\n")


def main():
    os.makedirs(COMPRESSED_DIR, exist_ok=True)
    os.makedirs(REFERENCE_DIR, exist_ok=True)

    fetch_silesia()
    fetch_nexrad()
    verify_all()
    print("Test dataset initialization and cryptographic validation complete!\n")


if __name__ == "__main__":
    main()

File: benchmarks/src/bin/fetch_data.rs
use std::fs::{self, File};
use std::io::{Read, Write};
use std::path::Path;

fn main() -> Result<(), Box<dyn std::error::Error>> {
    let compressed_dir = Path::new("data/compressed");
    let reference_dir = Path::new("data/reference");
    fs::create_dir_all(compressed_dir)?;
    fs::create_dir_all(reference_dir)?;

    println!("=== 1. Fetching NEXRAD Radar Test Data ===");
    let nexrad_urls = vec![
        ("https://unidata-nexrad-level2.s3.amazonaws.com/2024/01/01/KTLX/KTLX20240101_000004_V06", "nexrad1.bz2"),
        ("https://unidata-nexrad-level2.s3.amazonaws.com/2024/01/01/KTLX/KTLX20240101_000646_V06", "nexrad2.bz2"),
        ("https://unidata-nexrad-level2.s3.amazonaws.com/2024/01/01/KTLX/KTLX20240101_001328_V06", "nexrad3.bz2"),
    ];

    let client = reqwest::blocking::Client::builder()
        .user_agent("bzip2-benchmarks/1.0")
        .build()?;

    let mut base_payloads = Vec::new();
    for (url, filename) in nexrad_urls {
        let target_path = compressed_dir.join(filename);
        println!("Downloading NEXRAD: {} -> {:?}", url, target_path);
        let resp = client.get(url).send()?;
        let bytes = resp.bytes()?;
        let bz2_payload = if bytes.len() > 28 && &bytes[28..31] == b"BZh" {
            bytes[28..].to_vec()
        } else {
            bytes.to_vec()
        };
        fs::write(&target_path, &bz2_payload)?;
        base_payloads.push(bz2_payload);
    }

    // Expand to 30 NEXRAD files for a large dataset (~50 MB compressed)
    for i in 4..=30 {
        let base_bytes = &base_payloads[(i - 1) % base_payloads.len()];
        let target_path = compressed_dir.join(format!("nexrad{}.bz2", i));
        fs::write(&target_path, base_bytes)?;
    }

    println!("\n=== 2. Fetching Silesia Corpus Test Data ===");
    let silesia_files = vec![
        "dickens",
        "mozilla",
        "mr",
        "nci",
        "ooffice",
        "osdb",
        "reymont",
        "samba",
        "sao",
        "webster",
        "xml",
        "x-ray",
    ];

    for name in silesia_files {
        let raw_url = format!("https://raw.githubusercontent.com/yewq/Silesia-compression-corpus/main/{}", name);
        let target_bz2_path = compressed_dir.join(format!("silesia_{}.bz2", name));
        
        println!("Fetching Silesia file: {} -> {:?}", raw_url, target_bz2_path);
        let resp = client.get(&raw_url).send();
        let uncompressed_data = match resp {
            Ok(r) if r.status().is_success() => r.bytes()?.to_vec(),
            _ => {
                println!("Failed downloading {}, checking silesia-small.tar backup...", name);
                // Fallback to silesia-small.tar if offline/github raw fails
                extract_from_silesia_tar(name)?
            }
        };

        // Save reference uncompressed file
        let ref_path = reference_dir.join(format!("silesia_{}", name));
        fs::write(&ref_path, &uncompressed_data)?;

        // Compress using libbz2-rs-sys to get reference .bz2
        let bz2_data = compress_bz2(&uncompressed_data)?;
        fs::write(&target_bz2_path, &bz2_data)?;
        println!("Saved compressed {:?} ({} bytes raw -> {} bytes bz2)", target_bz2_path, uncompressed_data.len(), bz2_data.len());
    }

    println!("\n=== 3. Decompressing NEXRAD Data for Reference Output ===");
    for i in 1..=30 {
        let bz2_path = compressed_dir.join(format!("nexrad{}.bz2", i));
        let ref_path = reference_dir.join(format!("nexrad{}", i));
        let compressed_bytes = fs::read(&bz2_path)?;
        let decompressed_bytes = decompress_bz2(&compressed_bytes)?;
        fs::write(&ref_path, &decompressed_bytes)?;
        println!("Saved NEXRAD reference output {:?} ({} bytes uncompressed)", ref_path, decompressed_bytes.len());
    }

    println!("\nTest data setup complete!");
    Ok(())
}

fn extract_from_silesia_tar(filename: &str) -> Result<Vec<u8>, Box<dyn std::error::Error>> {
    let tar_path = Path::new("../libbzip2-rs/silesia-small.tar");
    let file = File::open(tar_path)?;
    let mut archive = tar::Archive::new(file);
    for entry in archive.entries()? {
        let mut entry = entry?;
        let path = entry.path()?;
        if path.to_string_lossy().ends_with(filename) {
            let mut data = Vec::new();
            entry.read_to_end(&mut data)?;
            return Ok(data);
        }
    }
    Err(format!("File {} not found in silesia-small.tar", filename).into())
}

fn compress_bz2(input: &[u8]) -> Result<Vec<u8>, Box<dyn std::error::Error>> {
    let mut dest = vec![0u8; input.len() + input.len() / 100 + 600];
    let mut dest_len = dest.len() as libc::c_uint;
    let res = unsafe {
        libbz2_rs_sys::BZ2_bzBuffToBuffCompress(
            dest.as_mut_ptr() as *mut libc::c_char,
            &mut dest_len,
            input.as_ptr() as *mut libc::c_char,
            input.len() as libc::c_uint,
            9,
            0,
            30,
        )
    };
    if res != libbz2_rs_sys::BZ_OK {
        return Err(format!("BZ2_bzBuffToBuffCompress failed with code {}", res).into());
    }
    dest.truncate(dest_len as usize);
    Ok(dest)
}

fn decompress_bz2(input: &[u8]) -> Result<Vec<u8>, Box<dyn std::error::Error>> {
    let mut out = Vec::new();
    let mut in_pos = 0;
    let mut is_first_stream = true;

    while in_pos < input.len() {
        // For subsequent streams in multi-stream archives (e.g. NEXRAD Level 2),
        // skip the 4-byte big-endian length header preceding each bzip2 stream.
        if !is_first_stream {
            if in_pos + 4 > input.len() {
                break;
            }
            let header = i32::from_be_bytes(input[in_pos..in_pos+4].try_into().unwrap());
            if header < 0 {
                break; // Negative header indicates end-of-volume marker
            }
            in_pos += 4;
        }

        // Verify BZh magic before attempting decompression
        if in_pos + 3 > input.len() || &input[in_pos..in_pos+3] != b"BZh" {
            break;
        }

        let mut stream: libbz2_rs_sys::bz_stream = unsafe { std::mem::zeroed() };
        let res = unsafe {
            libbz2_rs_sys::BZ2_bzDecompressInit(&mut stream, 0, 0)
        };
        if res != libbz2_rs_sys::BZ_OK {
            return Err(format!("BZ2_bzDecompressInit failed code {}", res).into());
        }

        loop {
            let chunk_in = &input[in_pos..];
            stream.next_in = chunk_in.as_ptr() as *mut libc::c_char;
            stream.avail_in = chunk_in.len() as libc::c_uint;

            let mut buf = [0u8; 64 * 1024];
            stream.next_out = buf.as_mut_ptr() as *mut libc::c_char;
            stream.avail_out = buf.len() as libc::c_uint;

            let ret = unsafe { libbz2_rs_sys::BZ2_bzDecompress(&mut stream) };
            let produced = buf.len() - stream.avail_out as usize;
            out.extend_from_slice(&buf[..produced]);

            let consumed = chunk_in.len() - stream.avail_in as usize;
            in_pos += consumed;

            if ret == libbz2_rs_sys::BZ_STREAM_END {
                break;
            } else if ret != libbz2_rs_sys::BZ_OK {
                unsafe { libbz2_rs_sys::BZ2_bzDecompressEnd(&mut stream) };
                return Err(format!("BZ2_bzDecompress returned error code {}", ret).into());
            }
        }

        unsafe { libbz2_rs_sys::BZ2_bzDecompressEnd(&mut stream) };
        is_first_stream = false;
    }

    Ok(out)
}
